modules/core/shared/src/main/scala/lucuma/core/util/Enumerated.scala

```diff
@@ -70,6 +72,23 @@ object Enumerated {
   def fromTag[A](implicit ev: Enumerated[A]): Prism[String, A] =
     Prism[String, A](ev.fromTag)(e => ev.tag(e))
 
+  private def enumValuesImpl[E: Type](using Quotes): Expr[Array[E]] =
+    import quotes.reflect.*
+    val companion = Ref(TypeRepr.of[E].typeSymbol.companionModule)
+    Select.unique(companion, "values").asExprOf[Array[E]]
+
+  private def tagImpl[E](x: Expr[E])(using Quotes): Expr[String] =
+    import quotes.reflect.*
+    Select.unique(x.asTerm, "tag").asExprOf[String]
+
+  private def enumeratedImpl[E: Type](using Quotes): Expr[Enumerated[E]] =
+    '{
+      Enumerated
+        .fromNEL(NonEmptyList.fromList(${ enumValuesImpl[E] }.toList).get)
+        .withTag(x => ${ tagImpl[E]('x) })
+    }
+
+  inline def derived[E]: Enumerated[E] = ${ enumeratedImpl[E] }
```

Comment:
Will this make the macro expand everywhere it is called? Or only once in the companion object?

Reply:
Technically, it is expanded wherever `Enumerated.derived` is called, which we shouldn't do more than once, or none at all: When `derives Enumerated` is used on an `enum`, this is syntactic sugar for using `Enumerated.derived` on its companion object.